File: androidApp/src/main/kotlin/com/chameleon/android/ui/viewmodel/NetworkGameViewModel.kt
package com.chameleon.android.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.chameleon.game.GameState
import com.chameleon.network.GameClient
import com.chameleon.network.GameServer
import com.chameleon.network.NetworkMessage
import com.chameleon.network.PlayerActionType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class NetworkGameUiState(
    val gameState: GameState = GameState(),
    val isHost: Boolean = false,
    val isConnected: Boolean = false,
    val connectionError: String? = null,
    val hostIp: String? = null,
    val currentPlayerId: String? = null,
    val isLoading: Boolean = false
)

class NetworkGameViewModel : ViewModel() {
    private val gameServer = GameServer()
    private val gameClient = GameClient()
    
    private val _uiState = MutableStateFlow(NetworkGameUiState())
    val uiState: StateFlow<NetworkGameUiState> = _uiState.asStateFlow()
    
    init {
        observeNetworkMessages()
        observeConnectionState()
        observeGameState()
    }
    
    private fun observeNetworkMessages() {
        viewModelScope.launch {
            gameClient.messageFlow.collect { message ->
                when (message) {
                    is NetworkMessage.JoinGameResponse -> {
                        if (message.success) {
                            _uiState.value = _uiState.value.copy(
                                currentPlayerId = message.playerId,
                                connectionError = null,
                                isLoading = false
                            )
                        } else {
                            _uiState.value = _uiState.value.copy(
                                connectionError = message.message,
                                isLoading = false
                            )
                        }
                    }
                    is NetworkMessage.GameStateUpdate -> {
                        _uiState.value = _uiState.value.copy(
                            gameState = message.gameState
                        )
                    }
                    is NetworkMessage.Error -> {
                        _uiState.value = _uiState.value.copy(
                            connectionError = message.message,
                            isLoading = false
                        )
                    }
                    else -> {}
                }
            }
        }
    }
    
    private fun observeConnectionState() {
        viewModelScope.launch {
            gameClient.connectionStateFlow.collect { isConnected ->
                _uiState.value = _uiState.value.copy(
                    isConnected = isConnected,
                    connectionError = if (!isConnected && !_uiState.value.isHost) "Connection lost" else null
                )
            }
        }
    }
    
    private fun observeGameState() {
        viewModelScope.launch {
            gameServer.gameStateFlow.collect { gameState ->
                if (_uiState.value.isHost) {
                    _uiState.value = _uiState.value.copy(
                        gameState = gameState
                    )
                }
            }
        }
    }
    
    fun startHosting(port: Int = 8080): Boolean {
        val hostIp = gameServer.startServer(port)
        return if (hostIp != null) {
            _uiState.value = _uiState.value.copy(
                isHost = true,
                isConnected = true,
                hostIp = hostIp,
                connectionError = null,
                currentPlayerId = "host_player"
            )
            
            viewModelScope.launch {
                gameClient.connect("localhost", port, "Host")
            }
            
            true
        } else {
            _uiState.value = _uiState.value.copy(
                connectionError = "Failed to start server"
            )
            false
        }
    }
    
    fun joinGame(hostIp: String, playerName: String) {
        _uiState.value = _uiState.value.copy(
            isLoading = true,
            connectionError = null
        )
        
        viewModelScope.launch {
            val success = gameClient.connect(hostIp, 8080, playerName)
            if (!success) {
                _uiState.value = _uiState.value.copy(
                    connectionError = "Failed to connect to host",
                    isLoading = false
                )
            }
        }
    }
    
    fun startGame(playerNames: List<String>) {
        if (_uiState.value.isHost) {
            gameServer.startGame(playerNames)
        }
    }
    
    fun submitClue(clue: String) {
        val playerId = _uiState.value.currentPlayerId ?: return
        
        viewModelScope.launch {
            gameClient.sendPlayerAction(playerId, PlayerActionType.SUBMIT_CLUE, clue)
        }
    }
    
    fun submitVote(votedPlayerId: String) {
        val playerId = _uiState.value.currentPlayerId ?: return
        
        viewModelScope.launch {
            gameClient.sendPlayerAction(playerId, PlayerActionType.SUBMIT_VOTE, votedPlayerId)
        }
    }
    
    fun submitChameleonGuess(guess: String) {
        val playerId = _uiState.value.currentPlayerId ?: return
        
        viewModelScope.launch {
            gameClient.sendPlayerAction(playerId, PlayerActionType.SUBMIT_CHAMELEON_GUESS, guess)
        }
    }
    
    fun startVoting() {
        val playerId = _uiState.value.currentPlayerId ?: return
        
        viewModelScope.launch {
            gameClient.sendPlayerAction(playerId, PlayerActionType.START_VOTING)
        }
    }
    
    fun disconnect() {
        viewModelScope.launch {
            if (_uiState.value.isHost) {
                gameServer.stopServer()
            } else {
                gameClient.disconnect()
            }
            
            _uiState.value = NetworkGameUiState()
        }
    }
    
    fun getSecretWordForPlayer(playerId: String): String? {
        return if (_uiState.value.gameState.chameleonPlayerId == playerId) {
            null
        } else {
            _uiState.value.gameState.secretWord
        }
    }
    
    override fun onCleared() {
        super.onCleared()
        disconnect()
    }
}

File: iosApp/iosApp/NetworkGameViewModel.swift
import Foundation
import Combine
import shared

@MainActor
class NetworkGameViewModel: ObservableObject {
    @Published var gameState = GameState()
    @Published var isConnected = false
    @Published var connectionError: String?
    @Published var backendIp = "192.168.1.100"
    @Published var backendPort = 8080
    @Published var currentPlayerId: String?
    @Published var isLoading = false
    
    private let gameClient = GameClient()
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        observeNetworkMessages()
        observeConnectionState()
    }
    
    private func observeNetworkMessages() {
        Task {
            for await message in gameClient.messageFlow {
                await handleNetworkMessage(message)
            }
        }
    }
    
    private func observeConnectionState() {
        Task {
            for await connected in gameClient.connectionStateFlow {
                await MainActor.run {
                    self.isConnected = connected
                    if !connected {
                        self.connectionError = "Connection lost"
                    }
                }
            }
        }
    }
    
    private func handleNetworkMessage(_ message: NetworkMessage) async {
        switch message {
        case let joinResponse as NetworkMessage.JoinGameResponse:
            if joinResponse.success {
                currentPlayerId = joinResponse.playerId
                connectionError = nil
                isLoading = false
            } else {
                connectionError = joinResponse.message
                isLoading = false
            }
        case let gameStateUpdate as NetworkMessage.GameStateUpdate:
            gameState = gameStateUpdate.gameState
        case let error as NetworkMessage.Error:
            connectionError = error.message
            isLoading = false
        default:
            break
        }
    }
    
    func createGame(playerName: String) {
        isLoading = true
        connectionError = nil
        
        Task {
            let success = try await gameClient.createGame(hostIp: backendIp, port: Int32(backendPort), playerName: playerName)
            await MainActor.run {
                if !success {
                    self.connectionError = "Failed to connect to backend server"
                    self.isLoading = false
                }
            }
        }
    }
    
    func setBackendAddress(ip: String, port: Int) {
        backendIp = ip
        backendPort = port
    }
    
    func joinGame(playerName: String) {
        isLoading = true
        connectionError = nil
        
        Task {
            let success = try await gameClient.connect(hostIp: backendIp, port: Int32(backendPort), playerName: playerName)
            await MainActor.run {
                if !success {
                    self.connectionError = "Failed to connect to backend server"
                    self.isLoading = false
                }
            }
        }
    }
    
    func startGame() {
        guard let playerId = currentPlayerId else { return }
        
        Task {
            try await gameClient.sendPlayerAction(playerId: playerId, action: PlayerActionType.readyForNextRound, data: "")
        }
    }
    
    func submitClue(_ clue: String) {
        guard let playerId = currentPlayerId else { return }
        
        Task {
            try await gameClient.sendPlayerAction(playerId: playerId, action: PlayerActionType.submitClue, data: clue)
        }
    }
    
    func submitVote(votedPlayerId: String) {
        guard let playerId = currentPlayerId else { return }
        
        Task {
            try await gameClient.sendPlayerAction(playerId: playerId, action: PlayerActionType.submitVote, data: votedPlayerId)
        }
    }
    
    func submitChameleonGuess(_ guess: String) {
        guard let playerId = currentPlayerId else { return }
        
        Task {
            try await gameClient.sendPlayerAction(playerId: playerId, action: PlayerActionType.submitChameleonGuess, data: guess)
        }
    }
    
    func startVoting() {
        guard let playerId = currentPlayerId else { return }
        
        Task {
            try await gameClient.sendPlayerAction(playerId: playerId, action: PlayerActionType.startVoting, data: "")
        }
    }
    
    func disconnect() {
        Task {
            try await gameClient.disconnect()
            await MainActor.run {
                self.gameState = GameState()
                self.isConnected = false
                self.connectionError = nil
                self.currentPlayerId = nil
                self.isLoading = false
            }
        }
    }
    
    func getSecretWordForPlayer(playerId: String) -> String? {
        return gameState.chameleonPlayerId == playerId ? nil : gameState.secretWord
    }
}
